### DISCUSSION TITLE
how-to-create-and-test-an-app-update-listener/

### DISCUSSION THREAD

**Discussion by giscus** [JakeSteam/blog-programming#13] [Comments]
# how-to-create-and-test-an-app-update-listener/

Running code when your app updates can be a useful marketing tool, and a reliable way of enabling new functionality only when the user updates.

https://blog.jakelee.co.uk/how-to-create-and-test-an-app-update-listener/

**Comment by BarryFruitman** [comment]
Really good, but I wouldn't suppress `@SuppressLint("UnsafeProtectedBroadcastReceiver")` since it protects your app, and the fix is even shorter than the suppression:

```
if (context == null || intent?action != "android.intent.action.MY_PACKAGE_REPLACED") {
```

**Reply by JakeSteam** [reply]
Good point! Have added it into the original post, thank you :)

**Comment by mciobanu** [comment]
Maybe I'm doing something wrong, but I'm afraid some updates are needed. My main concern is with the tests, as I didn't get to actually deploying anything from Google Play. I'll probably do that in some test variant.

I was trying to duplicate these in my app after finding similar suggestions elsewhere, and nothing happened. This post is clearer, though, so I tried to follow it, but it didn't help. Then I cloned the sample repo and there was still no toast on my Pixel 7 Pro. Then I started an Android 8.1 emulator and the thing finally worked, but that doesn't help me with Android 13. I also tested on a Huawei P20 Pro with Android 10, and there is a toast, just not the right one. It says *"updatelistener needs to be closed while it is updating"*.

Also, nitpick: I'd add a step for Gradle's "Sync now" after changing `versionCode`.
